File: R/conway.r
query <- function(g, x, y) {
    h <- dim(g)[1]
    w <- dim(g)[2]
    return(g[x %% h + 1, y %% w + 1])
}

assign <- function(g, x, y, state) {
    h <- dim(g)[1]
    w <- dim(g)[2]
    g[x %% h + 1, y %% w + 1] <- state
    return(g)
}

count_neighbors <- function(g, x, y) {
    n  <- query(g, x + 1, y    )
    ne <- query(g, x + 1, y + 1)
    e  <- query(g, x,     y + 1)
    se <- query(g, x - 1, y + 1)
    s  <- query(g, x - 1, y    )
    sw <- query(g, x - 1, y - 1)
    w  <- query(g, x    , y - 1)
    nw <- query(g, x + 1, y - 1)
    neighbor_states <- c(n, ne, e, se, s, sw, w, nw)
    return(sum(neighbor_states))
}

draw <- function(value) {
    if (value == 1) {
        return("#")
    } else {
        return(".")
    }
}

print_grid <- function(g) {
    for (r in seq(nrow(g))) {
        for (c in seq(ncol(g))) {
            cat(draw(g[r, c]))
        }
        cat("\n")
    }
}

game_logic <- function(state, neighbors) {
    if (state == 1) { 
        if (neighbors < 2) {
            return(0)
        }
        if (neighbors > 3) {
            return(0)
        }
        return(state)
    } else {
        if (neighbors == 3) {
            return(1)
        }
        return(state)
    }
}

step_cell <- function(g, x, y) {
    state <- query(g, x, y)
    neighbors <- count_neighbors(g, x, y)
    next_state <- game_logic(state, neighbors)
    return(next_state)
}

simulate <- function(g) {
    h <- dim(g)[1]
    w <- dim(g)[2]
    new_vec <- c(0)
    new_grid <- array(new_vec, c(h, w))
    for (r in seq(nrow(g))) {
        for (c in seq(ncol(g))) {
            new_state <- step_cell(g, r, c)
            new_grid <- assign(new_grid, r, c, new_state)
        }
    }
    return(new_grid)
}

clear_screen <- function() {
    cat("\x1b[2J")
}

height <- 10
width <- 20
vec <- c(0)
grid <- array(vec, c(height, width))

grid = assign(grid, 1, 4, 1)
grid = assign(grid, 2, 5, 1)
grid = assign(grid, 3, 3, 1)
grid = assign(grid, 3, 4, 1)
grid = assign(grid, 3, 5, 1)

print_grid(grid)

repeat {
    clear_screen()
    grid <- simulate(grid)
    print_grid(grid)
    Sys.sleep(0.1)
}

# Rscript conway.r

File: C#/Conway.cs
class Conway {
    private int height;
    private int width;
    private int[,] grid;

    Conway(int h, int w) {
        height = h;
        width = w;
        grid = new int[height, width];
    }

    private int Mod(int x, int y) {
        if (x > 0) {
            return x % y;
        }
        if (x < 0) {
            return y + x % y;
        }
        return 0;
    }

    private void Assign(int x, int y, int state) {
        grid[Mod(x, height), Mod(y, width)] = state;
    }

    private void Print() {
        for (int i = 0; i < grid.GetLength(0); i++) {
            for (int j = 0; j < grid.GetLength(1); j++) {
                Console.Write(grid[i, j] == 1 ? '#' : '.');
            }
            Console.WriteLine();
        }
    }

    private int Query(int x, int y) {
        return grid[Mod(x, height), Mod(y, width)];
    }

    private int CountNeighbors(int x, int y) {
        int n  = Query(x + 1, y    );
        int ne = Query(x + 1, y + 1);
        int e  = Query(x,     y + 1);
        int se = Query(x - 1, y + 1);
        int s  = Query(x - 1, y    );
        int sw = Query(x - 1, y - 1);
        int w  = Query(x,     y - 1);
        int nw = Query(x + 1, y - 1);
        int[] neighborStates = new int[] {n, ne, e, se, s, sw, w, nw};
        return neighborStates.Sum();
    }

    private int GameLogic(int state, int neighbors) {
        if (state == 1) {
            if (neighbors < 2) {
                return 0;
            }
            if (neighbors > 3) {
                return 0;
            }
            return state;
        } else {
            if (neighbors == 3) {
                return 1;
            }
            return state;
        }
    }

    private int StepCell(int x, int y) {
        int state = Query(x, y);
        int neighbors = CountNeighbors(x, y);
        int nextState = GameLogic(state, neighbors);
        return nextState;
    }

    private void Simulate() {
        int[,] newGrid = new int[height, width];
        for (int i = 0; i < grid.GetLength(0); i++) {
            for (int j = 0; j < grid.GetLength(1); j++) {
                newGrid[i, j] = StepCell(i, j);
            }
        }
        grid = newGrid;
    }

    private void ClearScreen() {
        Console.Write("\x1B[2J");
    }

    static void Main() {
        Conway con = new Conway(10, 20);
        con.Assign(0, 3, 1);
        con.Assign(1, 4, 1);
        con.Assign(2, 2, 1);
        con.Assign(2, 3, 1);
        con.Assign(2, 4, 1);

        con.Print();

        while (true) {
            con.ClearScreen();
            con.Simulate();
            con.Print();
            Thread.Sleep(100);
        }
    }
}

// dotnet Conway.cs
